File: concepts/01-arrays/node/arrayBasics.js
/**
 * Array Basics in Node.js
 * JavaScript arrays are dynamic arrays
 */

class DynamicArray {
    /**
     * Implementation of a dynamic array
     */
    constructor(capacity = 10) {
        this.capacity = capacity;
        this.size = 0;
        this.data = new Array(capacity).fill(null);
    }

    /**
     * Get length of array
     */
    get length() {
        return this.size;
    }

    /**
     * Get element at index - O(1)
     */
    get(index) {
        if (index >= 0 && index < this.size) {
            return this.data[index];
        }
        throw new Error("Index out of range");
    }

    /**
     * Set element at index - O(1)
     */
    set(index, value) {
        if (index >= 0 && index < this.size) {
            this.data[index] = value;
        } else {
            throw new Error("Index out of range");
        }
    }

    /**
     * Add element at the end - O(1) amortized
     */
    append(value) {
        if (this.size >= this.capacity) {
            this._resize();
        }
        this.data[this.size] = value;
        this.size++;
    }

    /**
     * Insert element at index - O(n)
     */
    insert(index, value) {
        if (index < 0 || index > this.size) {
            throw new Error("Index out of range");
        }

        if (this.size >= this.capacity) {
            this._resize();
        }

        // Shift elements to the right
        for (let i = this.size; i > index; i--) {
            this.data[i] = this.data[i - 1];
        }

        this.data[index] = value;
        this.size++;
    }

    /**
     * Delete element at index - O(n)
     */
    delete(index) {
        if (index < 0 || index >= this.size) {
            throw new Error("Index out of range");
        }

        // Shift elements to the left
        for (let i = index; i < this.size - 1; i++) {
            this.data[i] = this.data[i + 1];
        }

        this.size--;
        this.data[this.size] = null;
    }

    /**
     * Double the capacity - O(n)
     */
    _resize() {
        this.capacity *= 2;
        const newData = new Array(this.capacity).fill(null);
        for (let i = 0; i < this.size; i++) {
            newData[i] = this.data[i];
        }
        this.data = newData;
    }

    /**
     * Linear search - O(n)
     */
    search(value) {
        for (let i = 0; i < this.size; i++) {
            if (this.data[i] === value) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Display array elements
     */
    display() {
        return this.data.slice(0, this.size);
    }
}

// Example usage
if (require.main === module) {
    // Using JavaScript's built-in array (dynamic array)
    const arr = [1, 2, 3, 4, 5];
    console.log("Original array:", arr);

    // Access element - O(1)
    console.log("Element at index 2:", arr[2]);

    // Append - O(1) amortized
    arr.push(6);
    console.log("After push:", arr);

    // Insert - O(n)
    arr.splice(0, 0, 0);
    console.log("After insert at index 0:", arr);

    // Delete - O(n)
    const index = arr.indexOf(3);
    if (index > -1) {
        arr.splice(index, 1);
    }
    console.log("After remove 3:", arr);

    // Search - O(n)
    const searchIndex = arr.indexOf(5);
    console.log("Index of 5:", searchIndex);

    // Custom DynamicArray
    console.log("\n--- Custom DynamicArray ---");
    const dynArr = new DynamicArray();
    for (let i = 0; i < 5; i++) {
        dynArr.append(i * 2);
    }
    console.log("Array:", dynArr.display());
    console.log("Length:", dynArr.length);
    console.log("Element at index 2:", dynArr.get(2));

    dynArr.insert(2, 99);
    console.log("After insert at index 2:", dynArr.display());

    dynArr.delete(1);
    console.log("After delete at index 1:", dynArr.display());
}

module.exports = DynamicArray;

File: concepts/01-arrays/python/array_basics.py
"""
Array Basics in Python
Python lists are dynamic arrays
"""

class DynamicArray:
    """Implementation of a dynamic array"""
    
    def __init__(self, capacity=10):
        self.capacity = capacity
        self.size = 0
        self.data = [None] * capacity
    
    def __len__(self):
        return self.size
    
    def __getitem__(self, index):
        if 0 <= index < self.size:
            return self.data[index]
        raise IndexError("Index out of range")
    
    def __setitem__(self, index, value):
        if 0 <= index < self.size:
            self.data[index] = value
        else:
            raise IndexError("Index out of range")
    
    def append(self, value):
        """Add element at the end - O(1) amortized"""
        if self.size >= self.capacity:
            self._resize()
        self.data[self.size] = value
        self.size += 1
    
    def insert(self, index, value):
        """Insert element at index - O(n)"""
        if index < 0 or index > self.size:
            raise IndexError("Index out of range")
        
        if self.size >= self.capacity:
            self._resize()
        
        # Shift elements to the right
        for i in range(self.size, index, -1):
            self.data[i] = self.data[i - 1]
        
        self.data[index] = value
        self.size += 1
    
    def delete(self, index):
        """Delete element at index - O(n)"""
        if index < 0 or index >= self.size:
            raise IndexError("Index out of range")
        
        # Shift elements to the left
        for i in range(index, self.size - 1):
            self.data[i] = self.data[i + 1]
        
        self.size -= 1
        self.data[self.size] = None
    
    def _resize(self):
        """Double the capacity - O(n)"""
        self.capacity *= 2
        new_data = [None] * self.capacity
        for i in range(self.size):
            new_data[i] = self.data[i]
        self.data = new_data
    
    def search(self, value):
        """Linear search - O(n)"""
        for i in range(self.size):
            if self.data[i] == value:
                return i
        return -1
    
    def display(self):
        """Display array elements"""
        return [self.data[i] for i in range(self.size)]


# Example usage
if __name__ == "__main__":
    # Using Python's built-in list (dynamic array)
    arr = [1, 2, 3, 4, 5]
    print("Original array:", arr)
    
    # Access element - O(1)
    print("Element at index 2:", arr[2])
    
    # Append - O(1) amortized
    arr.append(6)
    print("After append:", arr)
    
    # Insert - O(n)
    arr.insert(0, 0)
    print("After insert at index 0:", arr)
    
    # Delete - O(n)
    arr.remove(3)
    print("After remove 3:", arr)
    
    # Search - O(n)
    index = arr.index(5)
    print("Index of 5:", index)
    
    # Custom DynamicArray
    print("\n--- Custom DynamicArray ---")
    dyn_arr = DynamicArray()
    for i in range(5):
        dyn_arr.append(i * 2)
    print("Array:", dyn_arr.display())
    print("Length:", len(dyn_arr))
    print("Element at index 2:", dyn_arr[2])
    
    dyn_arr.insert(2, 99)
    print("After insert at index 2:", dyn_arr.display())
    
    dyn_arr.delete(1)
    print("After delete at index 1:", dyn_arr.display())

